### PR TITLE
Fix run exports to restrict the minor version

### PR DESCRIPTION
**Pull request by vyasr**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

The current run exports for this library are too loose. Minor releases within the same major family are not ABI-compatible with one another. Arrow has never released a minor version before, so we've never observed this issue, but with the recent 16.1 release we see that the libraries packaged by this recipe all contain SOVERSIONs that are based on both the major and minor versions of the library, so packages built against one minor version cannot use a different one at runtime.

### CODE REVIEW

**Comment by vyasr** [comment]
@conda-forge-admin, please rerender

**Review by jakirkham — COMMENTED** [review]
Thanks Vyas! 🙏

Have a few questions below. Admittedly these may be things other maintainers have thoughts on as well

**Comment by h-vetinari** [comment]
> The current run exports for this library are too loose. Minor releases within the same major family are not ABI-compatible with one another. Arrow has never released a minor version before, so we've never observed this issue, but with the recent 16.1 release we see that the libraries packaged by this recipe all contain SOVERSIONs that are based on both the major and minor versions of the library, so packages built against one minor version cannot use a different one at runtime.

@raulcd, this is exactly what I meant when asking about arrow's first minor version bump in #1407. 😅 

Given that this was explicitly loosened in https://github.com/conda-forge/arrow-cpp-feedstock/issues/1096, perhaps the ABI _should_ be the same, and the libraries only versioned by the major versions? CC @kou @jorisvandenbossche

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml:120]: > In which case maybe a Jinja `for`-loop would be more DRY and easier to maintain

Let's not do that. It creates more comprehension overhead than it reduces

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml:113]: > Am noticing [in `pyarrow`](https://github.com/conda-forge/pyarrow-feedstock/blob/64ce0731d0a5e59c0e293c80b5ae7202d30ee021/recipe/meta.yaml#L89-L92), we are doing this

This is necessary because pyarrow is built on a different feedstock, and so cannot use `pin_subpackage`. 

> Do we want a similarly tight pin here or should `pyarrow` relax its pin?

Neither.

**Comment by h-vetinari** [comment]
@vyasr, could you send a PR to mark the just-released `libarrow*` builds for 16.1.0 as broken? I'd like to not have to rush this PR.

**Comment by vyasr** [comment]
Yup will do.

**Comment by vyasr** [comment]
Opened https://github.com/conda-forge/admin-requests/pull/996. Let me know if that looks right (there are a lot of arrow libraries, I went ahead and marked all of them).

**Comment by h-vetinari** [comment]
@jorisvandenbossche @kou @raulcd 
Can we please have some guidance on what arrow wants to do here?[^1] Should we change the pinning to minor level (meaning we have to remigrate again for 16.1 after 16.0), or should the minor level be removed from the SOVERSION (with the implication that the ABI doesn't change between minor versions)?

[^1]: There are several migrations waiting in the wings, which we cannot launch right now without reverting back to 16.0, as 16.1 is currently not publishable, unless we bite the bullet of this PR and then remigrate again.

**Comment by raulcd** [comment]
> should the minor level be removed from the SOVERSION (with the implication that the ABI doesn't change between minor versions)?

I have no idea why the original implementation did it this way. I wasn't aware the ABI would change between minor versions.

Nonetheless for 16.1.0 we already have a change of ABI so the only solution for conda is to pin the minor version. Otherwise we would require a new Arrow release and for conda to skip 16.1.0, right?

**Comment by h-vetinari** [comment]
> I have no idea why the original implementation did it this way.

Please see #1096 by your colleagues.

**Comment by h-vetinari** [comment]
> Nonetheless for 16.1.0 we already have a change of ABI

Do we know that the ABI actually changed? Or is it just the default way the SOVERSION is constructed (taking into account the minor version) that makes it _seem_ like the ABI has changed?

**Comment by jorisvandenbossche** [comment]
This is all very unclear and undocumented, a situation we should definitely improve on the Arrow side. 
But my understanding is that conda-forge actually should never have loosened those pins in the first place (I asked about it when the pin was loosened in https://github.com/apache/arrow/pull/37624#discussion_r1332748681 and https://github.com/conda-forge/arrow-cpp-feedstock/issues/1096#issuecomment-1670326019, but never got to follow up on that, my apologies).

See also https://github.com/apache/arrow/issues/41659 and https://github.com/apache/arrow/issues/41679

My understanding:

- Our release logic _clearly_ bumps the so version for minor feature releases (as we notice here), and this is done intentionally (https://github.com/apache/arrow/pull/4801, https://lists.apache.org/thread/r1ztq38xz5tn4p4v0xc1xk2d8x82p84v)
- We don't check for ABI stability, and given how easily you change the ABI in a C++ project, I doubt that our minor/features releases would be ABI stable (the 16.1.0 was a small feature release with only changes of two weeks, so that might "accidentally" have kept a stable ABI just because of only having few changes, but that's not am indication how future minor feature releases would be)
- The above point is actually a problem for our bug-fix releases, where we do promise a stable ABI, but are not very good at this .. (giving this, I would personally even consider going back to the most strict pinning at the patch level as it was originally, just to be safe)

**Comment by h-vetinari** [comment]
> Given this, I would personally even consider going back to the most strict pinning at the patch level as it was originally, just to be safe.

That's fair enough, though there's a cost to this safety - incompatibility of a lot of packages between each other, as well as the need to migrate all arrow-dependent packages every time. I'm OK with migrating per minor version if that's the reality, but I'd really ask if you can make it work to not break ABI on patch releases.

**Comment by jorisvandenbossche** [comment]
> I'm OK with migrating per minor version if that's the reality, but I'd really ask if you can make it work to not break ABI on patch releases.

Yep, that's fair. Migrating per minor version is what conda-forge should definitely do at the moment, because as long as we do not re-discuss and decide on this on the Arrow side, the official policy and what we do in practice is bumping the so version for minor versions. And we indeed should do a better job of ensuring to not break the ABI in patch releases (I also assume it hasn't happened much, I only remember this one report)

Anyway, I think starting with moving the pin to minor versions sounds good, and I opened the following issues to follow-up on the Arrow side:

- https://github.com/apache/arrow/issues/41707
- https://github.com/apache/arrow/issues/41708

**Comment by h-vetinari** [comment]
OK, sounds good to me, thanks. Before we can merge this PR, we need a repodata patch that tightens the run-deps of existing packages that built against libarrow 16 to depend on `>=16.0.0,<16.1` instead of `>=16.0.0,<17`. @vyasr, do you feel up for trying your hand at a repodata patch [here](https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/)?

**Comment by vyasr** [comment]
FWIW from the RAPIDS end libcudf has not encountered ABI-breaking issues with patch releases of arrow that I can recall. We have historically used a runtime pinning that allows both the minor and patch version to float, and it hasn't caused us any issues. The minor version of course hasn't mattered since arrow hasn't had a minor release before now, but at least that indicates that empirically patch versions haven't broken the ABI.

Sure I can put together a repodata patch. I won't get to it until next week, but I should be able to get it done early next week.

**Comment by h-vetinari** [comment]
OK thanks! In this case, I want to move forward with a bunch of migrations, so ended up doing it myself. 🙃 

If you're interested in how it's done (for next time 😉), see here: https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/731

**Review by h-vetinari — APPROVED** [review]
Thanks for the PR!

**Comment by vyasr** [comment]
@h-vetinari thanks! I'm familiar with repodata patches but I definitely would not have done it so fast.

@conda-forge/arrow-cpp we'll need to backport this to get new builds of 16.0 up that have the correct run exports. The fix here ensures that 16.1 builds are find, but any libraries that are building against 16.0 (which is still the conda-forge pinning) will still inherit the wrong run exports.

**Comment by vyasr** [comment]
Created an issue with this request in #1418

**Pull request merged**
